### PR TITLE
enh: frontend and openmp pass to support nested do loops and user defined functions in body

### LINKED ISSUE #4333: Compile parallel sparse matrix vector multiply with openmp
https://github.com/certik/hfsolver/blob/b4c50c1979fb7e468b1852b144ba756f5a51788d/src/sparse.f90#L200

```fortran
function csr_matvec(Ap, Aj, Ax, x) result(y)
! Compute y = A*x for CSR matrix A and dense vectors x, y
integer, intent(in) :: Ap(:), Aj(:)
real(dp), intent(in) :: Ax(:), x(:)
real(dp) :: y(size(Ap)-1)
integer :: i
!$omp parallel default(none) shared(Ap, Aj, Ax, x, y) private(i)
!$omp do
do i = 1, size(Ap)-1
    y(i) = dot_product(Ax(Ap(i):Ap(i+1)-1), x(Aj(Ap(i):Ap(i+1)-1)))
end do
!$omp end do
!$omp end parallel
end function
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
With this PR we get #4333 computed correctly. Although we use it's workaround but the workaround are for LFortran in general and not specific to openmp, I'll prefer to open issues for workaround and close it. Closes #4333. Towards #3777.

### CODE REVIEW

**Comment by Pranavchiku** [comment]
This is weird, gfortran test works for me:

```console
% ./integration_tests/run_tests.py -b gfortran
1172/1173 Test #1172: openmp_27 ................................   Passed    0.01 sec
1173/1173 Test #1163: openmp_18 ................................   Passed    0.48 sec

100% tests passed, 0 tests failed out of 1173

Label Time Summary:
c                 =   0.38 sec*proc (72 tests)
c_nopragma        =   0.02 sec*proc (3 tests)
cpp               =   0.23 sec*proc (45 tests)
fortran           =   1.50 sec*proc (329 tests)
gfortran          =   5.74 sec*proc (1173 tests)
llvm              =   4.44 sec*proc (992 tests)
llvm2             =   0.06 sec*proc (9 tests)
llvmImplicit      =   0.15 sec*proc (38 tests)
llvmStackArray    =   0.44 sec*proc (106 tests)
llvm_nopragma     =   0.02 sec*proc (3 tests)
llvm_omp          =   0.67 sec*proc (32 tests)
llvm_rtlib        =   0.01 sec*proc (1 test)
llvm_wasm         =   4.32 sec*proc (970 tests)
llvm_wasm_emcc    =   4.31 sec*proc (969 tests)
wasm              =   0.86 sec*proc (192 tests)
x86               =   0.10 sec*proc (19 tests)

Total Test time (real) =   1.22 sec
```

File: integration_tests/openmp_28.f90 [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+real, allocatable :: Ax(:)
+! Initialize CSR matrix A
+allocate(Ap(n+1), Aj(3*n), Ax(3*n))
+Ap = [(3*i, i=0,n)]
+Aj = [(3*i-1, 3*i, 3*i+1, i=1,n)]
+Ax = [(1.0, 2.0, 3.0, i=1,3*n)]
+! Initialize vector x
+x = [(i, i=1,n)]
+! Compute y = A*x
+call csr_matvec(Ap, Aj, Ax, x, y)
```
Comment (Pranavchiku) [inline integration_tests/openmp_28.f90:44]: I think the inputs are off. @certik is this the correct way to test it?

Comment (certik) [reply integration_tests/openmp_28.f90:44]: Yes, this is the correct way, we just need to ensure the Ap, Aj and Ax arrays are constructed correctly (no off by one errors).

Comment (certik) [reply integration_tests/openmp_28.f90:44]: There is an indexing bug:
```console
$ gfortran -g -fcheck=all  a.f90
$ ./a.out
At line 12 of file a.f90
Fortran runtime error: Index '0' of dimension 1 of array 'ax' below lower bound of 1

Error termination. Backtrace:
#0  0x562b3f86185b in csr_matvec_
        at /home/ondrej/repos/lfortran/lfortran/a.f90:12
#1  0x562b3f8623c7 in openmp_28
        at /home/ondrej/repos/lfortran/lfortran/a.f90:46
#2  0x562b3f862416 in main
        at /home/ondrej/repos/lfortran/lfortran/a.f90:22
```

Comment (Pranavchiku) [reply integration_tests/openmp_28.f90]: Yes, the latest commit fixed it.

**Pull request merged**